samples/connectedhomeip/light_bulb/src/AppTask.cpp

```diff
@@ -0,0 +1,345 @@
+		if (isFullyConnected) {
+			sStatusLED.Set(true);
+		} else if (sIsThreadProvisioned && sIsThreadEnabled && sIsPairedToAccount) {
+			sStatusLED.Blink(950, 50);
+		} else if (sHaveBLEConnections) {
+			sStatusLED.Blink(100, 100);
+		} else {
+			sStatusLED.Blink(50, 950);
+		}
+
+		sStatusLED.Animate();
+	}
+}
+
+void AppTask::PostEvent(const AppEvent& aEvent)
+{
+	if (k_msgq_put(&sAppEventQueue, &aEvent, K_NO_WAIT)) {
+		LOG_INF("Failed to post event to app task event queue");
+	}
+}
+
+void AppTask::UpdateClusterState()
+{
+	uint8_t onOff = LightingMgr().IsTurnedOn();
+
+	// write the new on/off value
+	EmberAfStatus status = emberAfWriteAttribute(1, ZCL_ON_OFF_CLUSTER_ID, ZCL_ON_OFF_ATTRIBUTE_ID, CLUSTER_MASK_SERVER, &onOff,
+						     ZCL_BOOLEAN_ATTRIBUTE_TYPE);
+
+	if (status != EMBER_ZCL_STATUS_SUCCESS) {
+		LOG_ERR("Updating on/off cluster failed: %x", status);
+	}
+
+	uint8_t level = LightingMgr().GetLevel();
+
+	status = emberAfWriteAttribute(1, ZCL_LEVEL_CONTROL_CLUSTER_ID, ZCL_CURRENT_LEVEL_ATTRIBUTE_ID, CLUSTER_MASK_SERVER, &level,
+					ZCL_DATA8_ATTRIBUTE_TYPE);
+
+	if (status != EMBER_ZCL_STATUS_SUCCESS) {
+		LOG_ERR("Updating level cluster failed: %x", status);
+	}
+}
+
+void AppTask::DispatchEvent(const AppEvent& aEvent)
+{
+	switch (aEvent.Type) {
+	case AppEvent::On:
+		LightingMgr().InitiateAction(LightingManager::Action::On, aEvent.Value, aEvent.LightEvent.ChipInitiated);
+		break;
+	case AppEvent::Off:
+		LightingMgr().InitiateAction(LightingManager::Action::Off, aEvent.Value, aEvent.LightEvent.ChipInitiated);
+		break;
+	case AppEvent::Toggle:
+		LightingMgr().InitiateAction(LightingMgr().IsTurnedOn() ? LightingManager::Action::Off : LightingManager::Action::On,
+					     aEvent.Value, aEvent.LightEvent.ChipInitiated);
+		break;
+	case AppEvent::Level:
+		LightingMgr().InitiateAction(LightingManager::Action::Level, aEvent.Value, aEvent.LightEvent.ChipInitiated);
+		break;
+	case AppEvent::FactoryReset:
+		LOG_INF("Factory Reset triggered");
+		ConfigurationMgr().InitiateFactoryReset();
+		break;
+	case AppEvent::StartThread:
+		StartThreadHandler();
+		break;
+	case AppEvent::StartBleAdvertising:
+		StartBLEAdvertisingHandler();
+		break;
+	case AppEvent::PublishLightBulbService:
+		sLightBulbPublishService.Publish();
+		break;
+	default:
+		LOG_INF("Unknown event received");
+		break;
+	}
+}
+
+void AppTask::ActionInitiated(LightingManager::Action aAction)
+{
+    if (aAction == LightingManager::Action::On)
+    {
+        LOG_INF("Turn On Action has been initiated");
+    }
+    else if (aAction == LightingManager::Action::Off)
+    {
+        LOG_INF("Turn Off Action has been initiated");
+    }
+    else if (aAction == LightingManager::Action::Level)
+    {
+        LOG_INF("Level Action has been initiated");
+    }
+}
+
+void AppTask::ActionCompleted(LightingManager::Action aAction)
+{
+    if (aAction == LightingManager::Action::On)
+    {
+        LOG_INF("Turn On Action has been completed");
```

Comment:
I was wondering on the same thing, but I see one use case of the current implementation. Imagine that someone is going to take this sample and try to expand somehow on his/her needs. In that case, he/she will be able to use this callbacks to perform some additional actions before starting and after ending action. If we will hide everything in the LightingManager it might be a bit harder to find proper place for writing own code. Does it make sense to you?

Reply:
Kind of yes :) Although if an action has an immediate effect, a caller of the action has means to perform additional actions before and after the main action by doing:
```
callAdditionalActionBeforeMain()
callMainAction()
callAdditionActionAfterMain()
```
so the callbacks seem to be a bit overkill to me, but I can imagine that they may be useful in case we need support actions that take longer time, like brightening the light slowly, so I'm OK with the current code. It was just a thought.